dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java

```diff
@@ -70,6 +70,30 @@
     private final LimitedChannelListener listener;
     private final ScheduledExecutorService scheduler;
     private final AtomicReference<BlacklistState> channelBlacklistState;
+    private final FutureCallback<Response> blacklistingCallback = new FutureCallback<Response>() {
+
+        @Override
+        public void onSuccess(Response response) {
+            // TODO(jellis): use the Retry-After header (if present) to determine how long to blacklist the channel
+            if (Responses.isQosStatus(response) || Responses.isServerError(response)) {
+                log.debug(
+                        "Blacklisting {} due to status code {}",
+                        UnsafeArg.of("delegate", delegate),
+                        SafeArg.of("code", response.code()));
+                channelBlacklistState.set(new BlacklistState(duration, NUM_PROBATION_REQUESTS));
```

Comment:
In cases where there are multiple requests in flight and they all fail, we will rest the blacklist multiple times. Should we CAS to start blacklisting on the first failure?

Reply:
I went back and forth on this, I think you're right though and we should avoid resetting the timeout when requests come in.

Reply:
Updated, it makes some of the simulations better :-)